lib/Segments/DynamicSegments/Filters/UserRole.php

```diff
@@ -25,15 +26,48 @@ public function apply(QueryBuilder $queryBuilder, DynamicSegmentFilterEntity $fi
     global $wpdb;
     $filterData = $filter->getFilterData();
     $role = $filterData->getParam('wordpressRole');
+    $operator = $filterData->getParam('operator');
     if (!$role) {
       throw new InvalidFilterException('Missing role', InvalidFilterException::MISSING_ROLE);
     }
+    if (!is_array($role)) {
+      // compatibility with the older segment before multiple roles were added
+      $role = [$role];
+    }
+    if (!$operator) {
+      $operator = DynamicSegmentFilterData::OPERATOR_ANY;
+    }
 
     $subscribersTable = $this->entityManager->getClassMetadata(SubscriberEntity::class)->getTableName();
-    $parameterSuffix = $filter->getId() ?? Security::generateRandomString();
-    return $queryBuilder->join($subscribersTable, $wpdb->users, 'wpusers', "$subscribersTable.wp_user_id = wpusers.id")
+    $parameterSuffix = ((string)$filter->getId()) . Security::generateRandomString();
```

Comment:
Out of curiosity, why are we using this? Is it to avoid hitting the cache?

Reply:
Hmm, I will try to explain, but it isn't straightforward, so please ask if you have more questions. Nope, it is not because of any cache. 

This is important when a customer connects multiple segments—we [pass the parameters](https://github.com/mailpoet/mailpoet/blob/master/lib/Segments/DynamicSegments/FilterHandler.php#L50) to the final query. But the customer can make a segment "(Subscribers in list A) OR (Subscribers in list B)". Then we need to pass the data "list A" and "list B" correctly to the final query. 

In other places, we have static parameter names ":segmentId," but that would pass it twice to the final query. Once for each sub-segment. And doctrine would only use "list A" in both queries, and the final query would look like: "(Subscribers in list A) OR (Subscribers in list A)". That means we need to make the parameters unique, so they are replaced correctly in the final big query. 

Does this make any sense?

Reply:
It does, thanks for explaining.